components/server/src/user/token-service.ts

```diff
@@ -63,42 +64,75 @@ export class TokenService implements TokenProvider {
         }
         const identity = this.getIdentityForHost(user, host);
 
+        function isValid(t: Token): boolean {
+            if (!t.expiryDate) {
+                return true;
+            }
+
+            const aboutToExpireTime = new Date();
+            aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
+            if (t.expiryDate >= aboutToExpireTime.toISOString()) {
+                return true;
+            }
+            return false;
+        }
+
         const doRefreshToken = async () => {
             // Check: Current token so we can actually refresh?
             const token = await this.userDB.findTokenForIdentity(identity);
             if (!token) {
+                reportScmTokenRefreshRequest(host, "no_token");
                 return undefined;
             }
 
-            const aboutToExpireTime = new Date();
-            aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
-            if (!token.expiryDate || token.expiryDate >= aboutToExpireTime.toISOString()) {
+            if (isValid(token)) {
+                reportScmTokenRefreshRequest(host, "still_valid");
                 return token;
             }
 
             // Can we refresh these kind of tokens?
             const { authProvider } = this.hostContextProvider.get(host)!;
             if (!authProvider.refreshToken) {
+                reportScmTokenRefreshRequest(host, "not_refreshable");
                 return undefined;
             }
 
-            await authProvider.refreshToken(user);
+            // Perform actual refresh
+            const stopTimer = scmTokenRefreshLatencyHistogram.startTimer({ host });
+            try {
+                await authProvider.refreshToken(user);
+            } finally {
+                stopTimer({ host });
+            }
+            reportScmTokenRefreshRequest(host, "success");
+
             return await this.userDB.findTokenForIdentity(identity);
         };
 
         try {
             const refreshedToken = await this.redisMutex.using(
                 [`token-refresh-${host}-${userId}`],
-                2000, // After 2s without extension the lock is released
+                3000, // After 3s without extension the lock is released
                 doRefreshToken,
-                { retryCount: 10, retryDelay: 500 }, // We wait at most 10s until we give up, and conclude that we can't refresh the token now.
+                { retryCount: 20, retryDelay: 500 }, // We wait at most 10s until we give up, and conclude that we can't refresh the token now.
```

Comment:
This was a mistake in the first PR: 10s was the intended value here.

In general, we did not have any timeout before. What we observe in prod is that these calls are taking ages (10s+), blocking UX in some cases. The metrics were added to better understand where that's coming from.